ai_edge_torch/generative/examples/gemma/gemma.py

```diff
@@ -36,16 +37,73 @@
     attn_value_proj="model.layers.{}.self_attn.v_proj",
     attn_output_proj="model.layers.{}.self_attn.o_proj",
     pre_attn_norm="model.layers.{}.input_layernorm",
-    pre_ff_norm="model.layers.{}.post_attention_layernorm",
+    post_attn_norm="model.layers.{}.post_attention_layernorm",
     embedding="model.embed_tokens",
     final_norm="model.norm",
     lm_head=None,
 )
 
+TENSOR_NAMES_GEMMA2 = loading_utils.ModelLoader.TensorNames(
+    ff_up_proj="model.layers.{}.mlp.up_proj",
+    ff_down_proj="model.layers.{}.mlp.down_proj",
+    ff_gate_proj="model.layers.{}.mlp.gate_proj",
+    attn_fused_qkv_proj="model.layers.{}.self_attn.qkv_proj",
+    attn_output_proj="model.layers.{}.self_attn.o_proj",
+    pre_attn_norm="model.layers.{}.input_layernorm",
+    post_attn_norm="model.layers.{}.post_attention_layernorm",
+    pre_ff_norm="model.layers.{}.pre_feedforward_layernorm",
+    post_ff_norm="model.layers.{}.post_feedforward_layernorm",
+    embedding="embedder",
+    final_norm="model.norm",
+    lm_head=None,
+)
+
+
+class Gemma2Block(TransformerBlock):
+
+  def forward(
+      self,
+      x: torch.Tensor,
+      rope: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
+      mask: Optional[torch.Tensor] = None,
+      input_pos: Optional[torch.Tensor] = None,
+  ) -> torch.Tensor:
+    """Forward function of the Gemma2Block.
+
+    Exactly the same as TransformerBlock but we call the post-attention norm
+    immediately after attention and not after the residual pointwise addition.
+
+    Args:
+      x (torch.Tensor): the input tensor.
+      rope (Tuple[torch.Tensor, torch.Tensor]): the input rope tensor.
+      mask (torch.Tensor): the optional mask tensor.
+      input_pos (torch.Tensor): the optional input position tensor.
+
+    Returns:
+      output activation from this transformer block.
+    """
+
+    if self.config.parallel_residual:
+      x_norm = self.pre_atten_norm(x)
+      attn_out = self.atten_func(x_norm, rope, mask, input_pos)
+      ff_out = self.ff(x_norm)
+      output = x + attn_out + ff_out
+    else:
+      x_norm = self.pre_atten_norm(x)
+      attn_out = self.atten_func(x_norm, rope, mask, input_pos)
+      attn_out_norm = self.post_attention_norm(attn_out)
+      x = x + attn_out_norm
+      output = x + self.ff(x)
+
+    return output
+
+
+T = TypeVar("T", TransformerBlock, Gemma2Block)
+
 
 class Gemma(nn.Module):
 
-  def __init__(self, config: cfg.ModelConfig):
+  def __init__(self, config: cfg.ModelConfig, block_type: T = TransformerBlock):
```

Comment:
oh i see, since you are reusing the logic here, it might actually save some code.

Reply:
Yes, I thought it would be reasonable to just config for Gemma2, but it's fine as a new file